### PR TITLE
Add November 2020 stress test doc

### PR DESCRIPTION
**Pull request by ajmaddaford**
### What is the context of this PR?
Documents the November 2020 stress test.

### How to review 
- Does the document accurately reflect the test?
- Are the observations accurate?
- Can any other observations be made?
- Do you agree with the recommendations?
- Should any other recommendations be made?

### CODE REVIEW

File: doc/performance-testing/0005_stress-test-november-2020.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+- The `loadbalancing.googleapis.com/https/backend_request_count` 500 metric as reported in Grafana (81 errors) didn't align with the errors reported by the load balancer (50 errors).
+
+## Recommendations
+
+- Update all environments to use threads and gRPC
+- Reduce the Redis instance size
+- Investigate GCS retries
+- Understand GCS scaling and speak to Google about the errors
+- Investigate the load profile we use to more evenly distribute writes to GCS
+- Test what happens when the application gets saturated; this could be done under lower load with a lower k8s node limit
```
Comment (pricem14pc) [inline doc/performance-testing/0005_stress-test-november-2020.md:121]: Does this mean pushing past the point that (as with this test) we see the latencies increase, until we observe actual errors?

Comment (ajmaddaford) [reply doc/performance-testing/0005_stress-test-november-2020.md:121]: Yes

File: doc/performance-testing/0005_stress-test-november-2020.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+- Redis only reached a peak of 11% CPU usage and 6,400 connected clients.
+- The error rate was very low. The errors happened in the middle of the test, for 4 of 20 test runs, after that there were 5 test runs with no errors. This may have been due to GCS scaling issues.
+- The previous Datastore errors did not occur, this looks to have been fixed by switching from HTTP to gRPC for Datastore calls.
+- Response times remained consistent until just over half of the allocated vCPU was used (900 of 1800 vCPU). Around this time the k8s cluster node scaling limit was reached.
+- 99th percentile response times reached 1 second at 95% CPU usage.
+- The `loadbalancing.googleapis.com/https/backend_request_count` 500 metric as reported in Grafana (81 errors) didn't align with the errors reported by the load balancer (50 errors).
+
+## Recommendations
+
+- Update all environments to use threads and gRPC
```
Comment (MebinAbraham) [inline doc/performance-testing/0005_stress-test-november-2020.md:116]: Given threads along with gRPC is very suitable, we should potentially look at uWSGI threads with GRPC if uWSGI is still a contender?

File: doc/performance-testing/0005_stress-test-november-2020.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        "requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))"
+
+## Observations
+
+- As sessions are stored in Datastore again, Redis was under much less load. This was the main contributing factor to a request rate of double the previous limit reached (74,000 rps vs 34,000 rps). 
+- Redis only reached a peak of 11% CPU usage and 6,400 connected clients.
+- The error rate was very low. The errors happened in the middle of the test, for 4 of 20 test runs, after that there were 5 test runs with no errors. This may have been due to GCS scaling issues.
+- The previous Datastore errors did not occur, this looks to have been fixed by switching from HTTP to gRPC for Datastore calls.
+- Response times remained consistent until just over half of the allocated vCPU was used (900 of 1800 vCPU). Around this time the k8s cluster node scaling limit was reached.
+- 99th percentile response times reached 1 second at 95% CPU usage.
```
Comment (MebinAbraham) [inline doc/performance-testing/0005_stress-test-november-2020.md:111]: I think it is also worth mentioning that while this is true, the 95th percentile response times were still less than 450ms at 95% CPU.

File: doc/performance-testing/0005_stress-test-november-2020.md (hunk 2 above)
Comment (ajmaddaford) [reply doc/performance-testing/0005_stress-test-november-2020.md:116]: Possibly, but I don't see that as an outcome of this test.

File: doc/performance-testing/0005_stress-test-november-2020.md (hunk 3 above)
Comment (ajmaddaford) [reply doc/performance-testing/0005_stress-test-november-2020.md]: Added 👍

File: doc/performance-testing/0005_stress-test-november-2020.md (hunk 2 above)
Comment (MebinAbraham) [reply doc/performance-testing/0005_stress-test-november-2020.md:116]: Yup, fair.

**Review by MebinAbraham — APPROVED** [review]
Observations ✅ 
Recommendations ✅

**Comment by pricem14pc** [comment]
I'm not sure it necessarily follows as a result of this performance test (although arguably this test has confirmed our proposed usage of Datastore over Redis) but I wonder if we should now capture a recommendation to test/resolve the eq-session accumulation now we have no automatic expiration?

**Comment by ajmaddaford** [comment]
> I'm not sure it necessarily follows as a result of this performance test (although arguably this test has confirmed our proposed usage of Datastore over Redis) but I wonder if we should now capture a recommendation to test/resolve the eq-session accumulation now we have no automatic expiration?

I don't think it should be added to the recommendations as I think the recommendations should be based on what was observed in the test. I have however added a card to the tech backlog to investigate options for purging sessions from Datastore.

**Review by pricem14pc — APPROVED** [review]

**Pull request merged**